### ISSUE TITLE
bug: false negative when upstream includes a function

### ISSUE THREAD

**Issue by NickvanDyke** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#53]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.7.0

### ESLint version

9.0.0

### Describe the bug

When analyzing a setter's args, the plugin collects references upstream to the args. e.g. if the arg comprises two other variables, the plugin considers said variables when determining if the setter arg is "internal" (React state or props).

These upstream references incidentally include function references - obviously never state. Thus the plugin considers the args to have some external component and never flags it. However the referenced function may only reference its parameters (pure) or state/props (internal), in which case it *should* be flagged. Only possible when the function is defined locally - not imported - for analysis.

Split from https://github.com/NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect/issues/34
Probably some interplay with https://github.com/NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect/issues/38

### Steps To Reproduce

Lint the following code:

```js
function computeNames(first, last) {
  return first + ' ' + last;
}

useEffect(() => {  
  // ❌ no error here  
  setNames(computeNames(firstName, lastName))  
  // ✅ this errors  
  setNames(firstName + lastName)  
}, [firstName, lastName])
```

No issue reported.

### Expected Behavior

Should report to avoid derived state on `setNames`.
